### PR TITLE
Save and load model state

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: samples/mnist/main.cpp [hunk 1]
```diff
@@ -36,14 +36,33 @@ int main(int argc, char *argv[]) {
...
-    // Temprorary buffers
+    deepworks::Tensor X, y;
+    dw::DataLoader val_loader(std::make_shared<custom::Dataset>(test_dir) , batch_size, /*shuffle */ false);
     deepworks::Tensor predict(model.outputs()[0].shape());
-    deepworks::Tensor grad_output(model.outputs()[0].shape());
-    deepworks::Tensor X_train, y_train;
 
-    dw::DataLoader train_loader(std::make_shared<custom::Dataset>(train_dir), batch_size, /*shuffle */ true);
-    dw::DataLoader val_loader  (std::make_shared<custom::Dataset>(test_dir) , batch_size, /*shuffle */ false);
+    // NB: If provided path to pretrained model just run validation.
```
Comment (TolyaTalamanov) [inline samples/mnist/main.cpp:43]: If path to pretrained model is provided

File: src/expression/nn.cpp [hunk 2]
```diff
@@ -8,89 +8,81 @@
...
-
-    // NB: Init beta.
-    deepworks::Tensor beta(deepworks::Shape{in_shape[1]});
-    deepworks::initializer::zeros(beta);
-    m_info.impl().params.emplace_back(std::move(beta), true);
+void dw::BatchNorm1D::init(const Shape& in_shape) {
+    // NB: Init trainable parameters and buffers.
+    m_info.impl().params.emplace ("gamma"       , dw::Tensor::constant({in_shape[1]}, 1.0));
+    m_info.impl().params.emplace ("beta"        , dw::Tensor::zeros({in_shape[1]}));
+    m_info.impl().buffers.emplace("running_mean", dw::Tensor::zeros({in_shape[1]}));
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/expression/nn.cpp:58]: So `running_mean` & `running_var` also should be stored to state, but they aren't parameters

File: src/model/model.cpp [hunk 3]
```diff
@@ -10,6 +10,7 @@
 #include "runtime/cpu/cpubackend.hpp"
 
 #include "util/assert.hpp"
+#include <iostream>
```
Comment (TolyaTalamanov) [inline src/model/model.cpp:13]: Remove

File: src/model/model.cpp [hunk 4]
```diff
@@ -128,15 +129,29 @@ deepworks::Model::Impl::Impl(deepworks::Placeholders ins,
                 m_layers.emplace_back(it->second);
 
                 // NB: Collect all parameters from every layer. (Used by optimizer)
-                for (auto&& p : it->second.params()) {
-                    m_params.emplace_back(p);
+                for (auto&& [name, p] : it->second.params()) {
```
Comment (TolyaTalamanov) [inline src/model/model.cpp:132]: p -> param
b -> buffer

File: src/model/model.cpp [hunk 5]
```diff
@@ -128,15 +129,29 @@ deepworks::Model::Impl::Impl(deepworks::Placeholders ins,
...
+                    auto it = m_params.emplace(info.name() + std::string(".") + name, p).first;
+                    m_state.emplace(it->first, p.data());
+                }
+
+                // NB: Push named buffers to state.
+                for (auto&& [name, b] : it->second.buffers()) {
+                    m_state.emplace(info.name() + std::string(".") + name, b);
```
Comment (TolyaTalamanov) [inline src/model/model.cpp:139]: Example: `linear0.bias`

File: tests/unit/test_networks.cpp [hunk 6]
```diff
@@ -344,6 +339,16 @@ TEST_F(MNISTModel, TrainLoopSmoke) {
         }
     }
 
+    //dw::save(model.state(), "model_state.bin");
```
Comment (TolyaTalamanov) [inline tests/unit/test_networks.cpp:342]: Remove

File: tests/unit/test_serialization.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    int batch_size   = 1;
+    int in_features  = 10;
+    int out_features = 2;
+};
+
+TEST_F(SimpleModelSerializationTest, SaveStateDict) {
+    auto m1 = buildModel();
+    dw::save(m1.state(), "state.bin");
+
+    for (const auto& [name, tensor] : m1.state()) {
```
Comment (TolyaTalamanov) [inline tests/unit/test_serialization.cpp:25]: Remove

**Comment by TolyaTalamanov** [comment]
@RodionovF @l-bat @KSTconsult @apertovs Have a look, please

File: tests/unit/test_cpulinear.cpp [hunk 8]
```diff
@@ -76,10 +76,10 @@ struct CPULinearAfterConvolutionModelTest : public ::testing::Test {
               model(buildModel()) {
         model.compile();
 
-        W_conv  = model.getLayer("conv1").params()[0].data();
-        b_conv  = model.getLayer("conv1").params()[1].data();
-        W       = model.getLayer("linear2").params()[0].data();
-        b       = model.getLayer("linear2").params()[1].data();
+	    W_conv  = model.getLayer("conv1").params().at("weight").data();
```
Comment (RodionovF) [inline tests/unit/test_cpulinear.cpp:79]: extra tab

Comment (TolyaTalamanov) [reply tests/unit/test_cpulinear.cpp]: Done, thanks!

File: samples/mnist/main.cpp [hunk 9]
```diff
@@ -36,14 +36,33 @@ int main(int argc, char *argv[]) {
...
+        int val_iter = 0;
+
+        // NB: Validation loop:
+        while (val_loader.pull(X, y)) {
+            model.forward(X, predict);
+            acc += dw::metric::accuracy(predict, y);
+            ++val_iter;
+        }
+
+        acc /= val_iter;
```
Comment (l-bat) [inline samples/mnist/main.cpp:58]: Why not?
```suggestion
        acc /= val_iter * batch_size;
```

**Review by l-bat — COMMENTED** [review]
Great! Please add more tests

File: samples/mnist/main.cpp (hunk 9 above)
Comment (TolyaTalamanov) [reply samples/mnist/main.cpp:58]: Because it's already done in accuracy ?

**Review by l-bat — APPROVED** [review]
